### PR TITLE
feat: update react & react-dom to v17

### PR DESCRIPTION
**Pull request by Mashal-m**
# Ticket
[Upgrade React JS to v17](https://github.com/openedx/frontend-app-admin-portal/issues/1000)

# Description 

- Updated `react` & `react-dom` to v17 along with `react-test-renderer`, `react-helmet`, `react-textarea-autosize`, `@testing-library/react` and `react-redux` to respective compatible versions
- Since Enzyme does not provide support React 17 ([enzymejs/enzyme#2429](https://github.com/enzymejs/enzyme/issues/2429)) or 18 ([enzymejs/enzyme#2524](https://github.com/enzymejs/enzyme/issues/2524)) itself we switched to community-supported project that aim to provide compatibility between Enzyme and React 17, replacing `enzyme-adapter-react-16` with `@wojtekmaj/enzyme-adapter-react-17` 
- Replaced `react-truncate` with `react-lines-ellipsis` due to dependency issues with react v17
- Update edx packages `@edx/frontend-enterprise-catalog-search`, `@edx/frontend-enterprise-hotjar` , `@edx/frontend-enterprise-logistration` `@edx/frontend-enterprise-utils` `@edx/frontend-platform`, and `@edx/paragon`

# For all changes

- [ ] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [ ] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

### CODE REVIEW

File: src/components/ContentHighlights/ContentHighlightCardItem.jsx [hunk 1]
```diff
@@ -19,14 +19,24 @@ const ContentHighlightCardItem = ({
     cardImgSrc: cardImageUrl,
     cardLogoSrc: partners.length === 1 ? partners[0].logoImageUrl : undefined,
     cardLogoAlt: partners.length === 1 ? `${partners[0].name}'s logo` : undefined,
-    cardTitle: <Truncate lines={3} title={title}>{title}</Truncate>,
+    cardTitle: <LinesEllipsis
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/ContentHighlightCardItem.jsx:22]: [curious] Why the change from `react-truncate`? 

Could Paragon's `Truncate` component work here? See documentation: https://paragon-openedx.netlify.app/components/truncate/

File: src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx [hunk 2]
```diff
@@ -154,6 +156,8 @@ describe('ContentHighlightCatalogVisibilityRadioInput1', () => {
     }).then(data => data));
 
     expect(EnterpriseCatalogApiService.updateEnterpriseCurationConfig).toHaveBeenCalledTimes(1);
-    expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(1);
+    /* Upgrading the @edx/paragon version from 20.41.0 to 20.42.0
+    caused this function to be called twice. Setting this to 2 in order to fix the test */
+    expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(2);
```
Comment (adamstankiewicz) [inline src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx:161]: Hmm, this doesn't feel quite like a "fix"... If the component itself is calling `sendEnterpriseTrackEvent` more than expected when the use toggled between these radio options (see screenshot), then that is a bug as it would skew the product analytics around this event (duplicate events for a single click).

Can this be reproduced on the [Paragon component](https://paragon-openedx.netlify.app/components/form/form-radio/) itself (i.e., `onChange` callback function passed to `Form.Radioset`)? If not, then it might be an issue with this repo's usage of the Paragon component.

image

File: src/components/settings/SettingsLMSTab/tests/AuthorizationsConfigs.test.tsx [hunk 3]
```diff
@@ -242,38 +242,40 @@ describe('Test authorization flows for Blackboard and Canvas', () => {
     expect(mockCanvasFetch).toHaveBeenCalledWith(1);
   });
   test('Canvas config is activated after last step', async () => {
-    renderWithRouter(<SettingsCanvasWrapper />);
-    mockCanvasUpdate.mockResolvedValue({ data: mockCanvasResponseDataActive });
+    setTimeout(async () => {
```
Comment (adamstankiewicz) [inline src/components/settings/SettingsLMSTab/tests/AuthorizationsConfigs.test.tsx:245]: [curious] Why is a `setTimeout` necessary here? It feels like a code smell that `setTimeout` would be needed in a test like this when its used React Testing Library.

File: package.json [hunk 4]
```diff
@@ -27,12 +27,13 @@
...
-    "@edx/frontend-enterprise-utils": "3.2.0",
-    "@edx/frontend-platform": "4.0.1",
-    "@edx/paragon": "20.39.2",
+    "@edx/frontend-enterprise-catalog-search": "^4.5.0",
+    "@edx/frontend-enterprise-hotjar": "^1.4.0",
+    "@edx/frontend-enterprise-logistration": "^3.4.0",
+    "@edx/frontend-enterprise-utils": "^3.4.0",
+    "@edx/frontend-platform": "^4.4.0",
+    "@edx/paragon": "^20.45.5",
+    "@wojtekmaj/enzyme-adapter-react-17": "^0.8.0",
```
[the remark this answers is not in the document]
Comment (Mashal-m) [reply package.json]: sure, I have pinned.

File: src/components/ContentHighlights/tests/ContentHighlightsCardItemsContainer.test.jsx [hunk 5]
```diff
@@ -42,8 +42,8 @@ describe('<ContentHighlightsCardItemsContainer>', () => {
     />);
     const firstTitle = testHighlightSet[0].title;
     const lastTitle = testHighlightSet[testHighlightSet.length - 1].title;
-    expect(screen.getByText(firstTitle)).toBeInTheDocument();
-    expect(screen.getByText(lastTitle)).toBeInTheDocument();
+    waitFor(() => expect(screen.getByText(firstTitle)).toBeInTheDocument());
+    waitFor(() => expect(expect(screen.getByText(lastTitle)).toBeInTheDocument()));
```
[the remark this answers is not in the document]
Comment (Mashal-m) [reply src/components/ContentHighlights/tests/ContentHighlightsCardItemsContainer.test.jsx]: That was by mistake, fixed. Thank you.

File: src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx (hunk 2 above)
Comment (Mashal-m) [reply src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx:161]: I've implemented the Form.Radioset component in another part of my code, and I'm facing an issue where the onChange function is getting called twice.

Attaching the code and UI snippet below. 
Screenshot 2023-08-24 at 6 23 11 PM
Screenshot 2023-08-24 at 6 26 16 PM

File: src/components/settings/SettingsLMSTab/tests/AuthorizationsConfigs.test.tsx (hunk 3 above)
Comment (Mashal-m) [reply src/components/settings/SettingsLMSTab/tests/AuthorizationsConfigs.test.tsx]: The initial issue was I ran into a "Null: create event" error when running this test file. The problem seemed to be with the last section. Async/await operations weren't working properly there, so I had to use setTimeout to add a small delay in order to clear the current stack.

I have made some changes for this particular test case ([commit](https://github.com/openedx/frontend-app-admin-portal/pull/1009#:~:text=authorization%20test%20file-,db7fdce,-Mashal%2Dm%20force))

File: src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx (hunk 2 above)
Comment (adamstankiewicz) [reply src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx:161]: @Mashal-m I believe this is related to [this Paragon bug](https://github.com/openedx/paragon/issues/2683) and [associated fix](https://github.com/openedx/paragon/pull/2705) (in code review), and will apply to Paragon v21. 

This same issue came up recently when trying to upgrade Paragon to latest v20 as well. We opted to defer on the Paragon latest v20 upgrade as this regression ends up causing duplicate success `Toast` messages to appear within some of our features in this MFE...

[question] Is it possible to move forward with this PR _without_ upgrading Paragon to latest v20 to avoid introducing this regression?

Comment (Mashal-m) [reply src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx:161]: `npm start` is failing on the master branch, and I'm encountering the following error (attached screenshot). The fix for this error was released in [v20.43.3](https://github.com/openedx/paragon/releases?page=5#:~:text=Compare-,v20.43.3,-20.43.3%20(2023%2D06 ). However, [v20.42.0](https://github.com/openedx/paragon/releases?page=6#:~:text=Compare-,v20.42.0,-20.42.0%20(2023%2D06 ) is causing this test failure, so to resolve the master issue, we need to upgrade the Paragon version.
Screenshot 2023-10-13 at 12 14 44 PM

Comment (adamstankiewicz) [reply src/components/ContentHighlights/CatalogVisibility/tests/ContentHighlightCatalogVisibilityRadioInput.test.jsx:161]: Ah, got it. Thanks for clarifying. Luckily, the aforementioned bug preventing us from upgrading to latest v20, causing the [issue](https://github.com/openedx/paragon/issues/2683) with the duplicate `onChange` calls was [fixed and backported](https://github.com/openedx/paragon/issues/2683#issuecomment-1761224904) to v20.46.3 earlier today. 

We should be able to move forward with upgrading to latest v20 at this point, without needing to update tests regarding `expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(2);` 🤞

**Review by adamstankiewicz — COMMENTED** [review]
Left some additional minor, non-blocking comments. Once Paragon upgrades to v20.46.3 to pull in the fix for the duplicate `onChange` in a couple Paragon components, this PR should be good to go.

File: package.json [hunk 6]
```diff
@@ -50,16 +51,15 @@
     "lodash": "4.17.21",
     "lodash.debounce": "4.0.8",
     "prop-types": "15.7.2",
-    "react": "16.14.0",
-    "react-dom": "16.13.1",
-    "react-helmet": "5.2.1",
+    "react": "^17.0.2",
```
[the remark this answers is not in the document]
Comment (Mashal-m) [reply package.json]: I have pinned it.

File: package.json [hunk 7]
```diff
@@ -27,12 +27,13 @@
...
-    "@edx/frontend-enterprise-utils": "3.2.0",
-    "@edx/frontend-platform": "4.0.1",
-    "@edx/paragon": "20.39.2",
+    "@edx/frontend-enterprise-catalog-search": "4.5.0",
+    "@edx/frontend-enterprise-hotjar": "1.4.0",
+    "@edx/frontend-enterprise-logistration": "3.4.0",
+    "@edx/frontend-enterprise-utils": "3.4.0",
+    "@edx/frontend-platform": "4.4.0",
+    "@edx/paragon": "20.45.5",
+    "@wojtekmaj/enzyme-adapter-react-17": "0.8.0",
```
[the remark this answers is not in the document]
Comment (Mashal-m) [reply package.json]: yes, correct.

File: src/components/ContentHighlights/tests/ContentHighlightsCardItemsContainer.test.jsx [hunk 8]
```diff
@@ -42,8 +42,8 @@ describe('<ContentHighlightsCardItemsContainer>', () => {
     />);
     const firstTitle = testHighlightSet[0].title;
     const lastTitle = testHighlightSet[testHighlightSet.length - 1].title;
-    expect(screen.getByText(firstTitle)).toBeInTheDocument();
-    expect(screen.getByText(lastTitle)).toBeInTheDocument();
+    waitFor(() => expect(screen.getByText(firstTitle)).toBeInTheDocument());
+    waitFor(() => expect(screen.getByText(lastTitle)).toBeInTheDocument());
```
[the remark this answers is not in the document]
Comment (Mashal-m) [reply src/components/ContentHighlights/tests/ContentHighlightsCardItemsContainer.test.jsx]: I have checked again, tests are passing without `waitFor`

**Comment by Mashal-m** [comment]
> Left some additional minor, non-blocking comments. Once Paragon upgrades to v20.46.3 to pull in the fix for the duplicate `onChange` in a couple Paragon components, this PR should be good to go.

I have pushed the requested changes and upgraded Paragon as well, we are good to go!

File: package.json [hunk 9]
```diff
@@ -56,15 +56,14 @@
     "lodash": "4.17.21",
     "lodash.debounce": "4.0.8",
     "prop-types": "15.7.2",
-    "react": "16.14.0",
-    "react-dom": "16.13.1",
-    "react-helmet": "5.2.1",
+    "react": "17.0.2",
+    "react-dom": "17.0.2",
+    "react-helmet": "^6.1.0",
```
Comment (brobro10000) [inline package.json:61]: [nit] I would pin this upgrade 👍🏽

**Review by brobro10000 — COMMENTED** [review]
I did notice some failing tests in the PR. 
For the `EmailAddressTableCell.tests.jsx` failing tests, refactoring to the await waitFor syntax seemed to fix the issue with the enterprise track event count and could be used to fix up similar issues the tests are showing
```js
    await waitFor(() => expect(screen.findByText('Learner data disabled', { exact: false })));
    await waitFor(() => expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(1));

```

Great job refactoring to paragons `Truncate` component 👍🏽

**Comment by Mashal-m** [comment]
> I did notice some failing tests in the PR. For the `EmailAddressTableCell.tests.jsx` failing tests, refactoring to the await waitFor syntax seemed to fix the issue with the enterprise track event count and could be used to fix up similar issues the tests are showing
> 
> ```js
>     await waitFor(() => expect(screen.findByText('Learner data disabled', { exact: false })));
>     await waitFor(() => expect(sendEnterpriseTrackEvent).toHaveBeenCalledTimes(1));
> ```
> 
> Great job refactoring to paragons `Truncate` component 👍🏽

Fixed the tests, it is ready for merge now.

**Review by brobro10000 — APPROVED** [review]
